jaspResults rewrite - Correlation

Kucharssim opened the pull request:
On top of that, adds some functionality:

- reporting sample size: https://github.com/jasp-stats/jasp-issues/issues/528 and https://github.com/jasp-stats/jasp-issues/issues/245
- choice between list-wise/pair-wise exclusion: https://github.com/jasp-stats/jasp-issues/issues/244
- pairwise plots (although seeing Don's improvement of this in descriptives this was probably overkill :D): https://github.com/jasp-stats/jasp-issues/issues/298
- assumption checks: https://github.com/jasp-stats/jasp-issues/issues/488
- add correlation heatmap plot (cannot find the issue but I am pretty sure @EJWagenmakers asked me to do this like a year ago :D)

- prepared some structure for partial correlations but this is still only in debug mode since we need to do some research on non-parametric confidence intervals.

Kucharssim committed:
implement pairwise table

Kucharssim committed:
use pairwise state

Kucharssim committed:
fix some dependencies

Kucharssim committed:
replace unnecesary call purrr::transpose with lapply

Kucharssim committed:
implement correlation table

Kucharssim committed:
improvement of correlation table

Kucharssim committed:
matrix plot

Kucharssim committed:
improve correlation matrix plot

Kucharssim committed:
add dependency

Kucharssim committed:
add pairwise plots

Kucharssim committed:
add flagging significant

Kucharssim committed:
add sample size

Kucharssim committed:
add assumption checks

Kucharssim committed:
init partial correlations

Kucharssim committed:
fix errors

Kucharssim committed:
change some labels in ui and improve assumptions

Kucharssim committed:
add heatmaps

Kucharssim committed:
clean up

Kucharssim committed:
more robust plots

Kucharssim committed:
fix small bug

Kucharssim committed:
more robust tables

Kucharssim committed:
robustify plots

Kucharssim committed:
small fixes

Kucharssim committed:
remove unnecessary error checks

Kucharssim committed:
update documentation

Kucharssim committed:
final touches

Kucharssim committed:
remove old .plotCorrelations function

Kucharssim committed:
add unit tests

TimKDJ commented on Resources/Regression/qml/Correlation.qml:
```diff
@@ -23,13 +23,14 @@ import JASP.Theme 1.0
 Form
 {
-	usesJaspResults: false
+    usesJaspResults: true
```
This line can go

TimKDJ commented on Resources/Regression/qml/Correlation.qml:
```diff
@@ -23,13 +23,14 @@ import JASP.Theme 1.0
 		height: Theme.smallDefaultVariablesFormHeight
 		AvailableVariablesList { name: "allVariablesList" }
 		AssignedVariablesList { name: "variables"; title: qsTr("Variables"); suggestedColumns: ["ordinal", "scale"] }
+        AssignedVariablesList { name: "conditioningVariables"; title: qsTr("Condition on"); suggestedColumns: ["ordinal", "scale"]; debug: true }
```
Qml uses tabs :)

TimKDJ commented on JASP-Engine/JASP/R/correlation.R:
```diff
@@ -14,644 +14,961 @@
+  if(!is.null(jaspResults[['assumptionsContainer']])) return()
+  
+  assumptionsContainer <- createJaspContainer(title = "Assumption checks")
+  assumptionsContainer$dependOn(c("multivariateShapiro", "pairwiseShapiro",
```
If you set `"multivariateShapiro", "pairwiseShapiro"` on their respective tables instead of the container, then you don't recompute multivariate when you toggle pairwise and vice versa. You'd have to check for the existence of the tables tho

Kucharssim replied:
Yeah, thanks for the suggestion!

TimKDJ commented on JASP-Engine/JASP/R/correlation.R:
```diff
@@ -14,644 +14,961 @@
+          }
+          shapiroTable$addRows(list(vars = names(shapiroResult)[i],
+                                    W    = shapiroResult[[i]]$result$statistic,
+                                    p    = shapiroResult[[i]]$result$p.value))
```
This would go nicely into a function of its own (same for pairwiseShapiro)

Kucharssim replied:
I moved these parts into their own functions

TimKDJ commented on JASP-Engine/JASP/R/correlation.R:
```diff
@@ -14,644 +14,961 @@
+  } else{
+    heatmaps <- createJaspContainer(title = "Heatmaps")
+    heatmaps$dependOn(c("variables", "conditioningVariables", "missingValues", "pearson", "spearman", "kendallsTauB",
+                        "flagSignificant"))
```
Why does this have a dependency on "flagSignificant"?

Kucharssim replied:
because if flagSignificant == TRUE, the heatmap highlights significant correlations...

TimKDJ replied:
Ah okay, missed that

TimKDJ reviewed (changes requested):
Well done Simon! Cool new features too

Some minor code comments and I noticed the following during testing:
- [ ] Vovksellke has the same symbol as the significance indicator which is rather confusing
- [ ] We no longer have the footnote “all tests one-tailed, for positive correlation. Also the "* p < .05, etc" no longer has the “one-tailed” addition at the end. Not sure if we care about either, just want to make sure they're not forgotten
- [ ] You get additional stars under the table (in the screenshot 3, but depends on the hypothesis):
<img width="1264" alt="Screenshot 2019-10-31 at 14 23 51" src="https://user-images.githubusercontent.com/18737241/67958865-22740200-fbf8-11e9-8a4f-646c495dba61.png">

Kucharssim committed:
formatting fixes (VSMPR symbol, add note for one-sided tests), replace spaces with tabs in QML

Kucharssim committed:
multivariate and pairwise assumption checks now have their own functions and dependencies

Kucharssim commented:
Ooops, I should have squashed the commits I guess... Does it hurt this way?

Ok, I addressed all points raised, except this one:
> * [ ]  You get additional stars under the table (in the screenshot 3, but depends on the hypothesis):

Can I actually add a footnote without printing anything under the table? That is, I use addFootnote to add symbols *, **, and *** inside the table to flag the significant correlations, but then the symbol is also added below. I did not figure out how to do that; if it's not possible, should I make an issue for @JorisGoosen?

JorisGoosen commented:
@Kucharssim we can squash it here ;)

As far as the footnotes without footnotes go: I don't think this is supported right now. But if you make an issue for it we can add something like: "a footnote without a message is not listed below the table".

The point about the stars underneath the table is then solved right? 
And then we can squash+merge it.

TimKDJ commented on JASP-Engine/JASP/R/correlation.R:
```diff
@@ -107,7 +107,17 @@ Correlation <- function(jaspResults, dataset, options){
+    additionToFlagSignificant <- ", one-talied"
+  } else if(options[['hypothesis']] == "correlatedNegatively"){
+    mainTable$addFootnote(message = "All tests one-tailed, for negative correlation")
+    additionToFlagSignificant <- ", one-talied"
```
2x one-tailed instead of one-talied

Kucharssim replied:
oh christ :D

TimKDJ commented:
Just a typo fix and we're all good

Kucharssim committed:
fix typo
